pkg/registry/ociregistry.go

```diff
@@ -0,0 +1,164 @@
+		return fmt.Errorf("NewRegistry: %v", err)
+	}
+
+	tlsConfig := &tls.Config{
+		RootCAs:            certificates,
+		InsecureSkipVerify: or.Insecure,
+	}
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.TLSClientConfig = tlsConfig
+	authClient := &auth.Client{
+		Client: &http.Client{
+			Transport: transport,
+		},
+		Cache: auth.NewCache(),
+	}
+	authClient.SetUserAgent("eksa")
+	authClient.Credential = func(ctx context.Context, s string) (auth.Credential, error) {
+		return credentialStore.Credential(s)
+	}
+	or.registry.Client = authClient
+	return nil
+}
+
+// SetProject for registry destination.
+func (or *OCIRegistryClient) SetProject(project string) {
+	or.Project = project
+}
+
+// Destination of this storage registry.
+func (or *OCIRegistryClient) Destination(image releasev1.Image) string {
+	return strings.Replace(image.VersionedImage(), image.Registry()+"/", or.Host+"/"+or.Project, 1)
+}
+
+// GetReference gets digest or tag version.
+func (or *OCIRegistryClient) getCertificates() (certificates *x509.CertPool, err error) {
+	if len(or.CertFile) < 1 {
+		return nil, nil
+	}
+	fileContents, err := ioutil.ReadFile(or.CertFile)
+	if err != nil {
+		return nil, fmt.Errorf("error reading certificate file <%s>: %v", or.CertFile, err)
+	}
+	certPool := x509.NewCertPool()
+	certPool.AppendCertsFromPEM(fileContents)
+
+	return certPool, nil
+}
+
+// GetStorage object based on repository.
+func (or *OCIRegistryClient) GetStorage(ctx context.Context, image releasev1.Image) (repo orasregistry.Repository, err error) {
+	dstRepo := or.Project + image.Repository()
+	repo, err = or.OI.Repository(ctx, or.registry, dstRepo)
+	if err != nil {
+		return nil, fmt.Errorf("error creating repository %s: %v", dstRepo, err)
+	}
+	return repo, nil
+}
+
+// Copy a image from a source to a destination.
+func (or *OCIRegistryClient) Copy(ctx context.Context, image releasev1.Image, dstClient StorageClient) (err error) {
+	committed := &sync.Map{}
+	extendedCopyOptions := oras.DefaultExtendedCopyOptions
+	extendedCopyOptions.PostCopy = func(ctx context.Context, desc ocispec.Descriptor) error {
+		committed.Store(desc.Digest.String(), desc.Annotations[ocispec.AnnotationTitle])
+		return nil
+	}
+	extendedCopyOptions.OnCopySkipped = func(ctx context.Context, desc ocispec.Descriptor) error {
+		committed.Store(desc.Digest.String(), desc.Annotations[ocispec.AnnotationTitle])
+		return nil
+	}
+
+	srcStorage, err := or.GetStorage(ctx, image)
+	if err != nil {
+		return fmt.Errorf("registry copy source: %v", err)
+	}
+
+	var desc ocispec.Descriptor
+	or.registry.Reference.Reference = image.Version()
+	desc, err = or.OI.Resolve(ctx, srcStorage, or.registry.Reference.Reference)
+	if err != nil {
+		return fmt.Errorf("registry copy destination: %v", err)
+	}
+
+	dstStorage, err := dstClient.GetStorage(ctx, image)
+	if err != nil {
+		return fmt.Errorf("registry copy destination: %v", err)
+	}
+
+	fmt.Println(dstClient.Destination(image))
```

Comment:
could we use some kind of logger for this?

Reply:
Changed to use "log" 🤷